### ISSUE TITLE
Infinite recomposition when using accompanist's placeholder

### ISSUE THREAD

**Issue by mroczis** [coil-kt/coil#1088]
**Describe the bug**
I'm trying to integrate ghost loaders in the app using [Appcomanist's placeholder](https://google.github.io/accompanist/placeholder/) library. This library simply binds to `Modifier` and accepts property that tells if ghost loader should be visible. In order to know state I use `ImagePainter.state` property. It initially holds `State.Loading` value and after a while `State.Success` pops up. This causes recomposition and loader disappears. Since this moment recomposition occurs indefinitely. `State.Success` is kept but instance of `State.Success#painter` changes.

As I looked a bit into source code the problem seems to be in `ImagePainter#onRemembered()` which is called in a loop and does check if re-load is actually needed.

**To Reproduce**
I have caching enabled on `Application`'s level.
```
ImageLoader.Builder(applicationContext)
        .crossfade(true)
        .logger(DebugLogger())
        .okHttpClient {
            get<OkHttpClient>().newBuilder()
                .cache(CoilUtils.createDefaultCache(applicationContext))
                .build()
        }
        .build()
```

Then I simply load an image:
```
// You can use any image
val painter = rememberImagePainter("https://upload.wikimedia.org/wikipedia/commons/b/b6/Image_created_with_a_mobile_phone.png")

Image(
    painter = painter,
    contentDescription = null,
    modifier = Modifier.fillMaxSize().placeholder(
        visible = painter.state is ImagePainter.State.Loading,
        color = Color.Red
    )
)
```

**Logs/Screenshots**
```
RealImageLoader: 🧠 Successful (MEMORY_CACHE) - https://upload.wikimedia.org/wikipedia/commons/b/b6/Image_created_with_a_mobile_phone.png
RealImageLoader: ☁️  Successful (NETWORK) - https://upload.wikimedia.org/wikipedia/commons/b/b6/Image_created_with_a_mobile_phone.png
RealImageLoader: 🧠 Successful (MEMORY_CACHE) - https://upload.wikimedia.org/wikipedia/commons/b/b6/Image_created_with_a_mobile_phone.png
RealImageLoader: 🧠 Successful (MEMORY_CACHE) - https://upload.wikimedia.org/wikipedia/commons/b/b6/Image_created_with_a_mobile_phone.png
... prints indefinitely
```

**Version**
Coil: 1.4.0
Compose: 1.1.0-rc01
Appcompanist: 0.22.0-rc
Android version: any

**Comment by mroczis** [comment]
After some digging I found out that recomposition is caused by new instances of `ImageRequest` that are created by `rememberImagePainter()` method (coil.compose.ImagePainter.kt, line 72). `ImageRequest` instances should be also remembered to avoid infinite recomposition.

How to solve the issue:
```
val url = "https://..."; // Insert whatever here
val request = remember { ImageRequest.Builder(context).data(url).build() }
val painter = rememberImagePainter(request)
```

**Comment by colinrtwhite** [comment]
Thanks for digging in - are you setting a `listener` or another anonymous function property that could change `ImageRequest`'s `hashCode` on each iteration?

**Comment by mroczis** [comment]
Yes, you are right. I checked `hashCode` and `toString` methods of `ImageRequest` and both of them return same values all the time.

**Comment by colinrtwhite** [comment]
@mroczis Just tested locally and this is fixed in `2.0.0-rc01`. Though one thing to note is `painter.state` is always `Loading` for the first composition so the placeholder animation runs even if the image is in the memory cache. Check out these docs for how to account for that: https://coil-kt.github.io/coil/compose/#observing-asyncimagepainterstate
